aws/sdk-codegen/src/test/kotlin/software/amazon/smithy/rustsdk/CredentialCacheConfigTest.kt

```diff
@@ -0,0 +1,150 @@
+                        let sut = ConfigOverrideRuntimePlugin {
+                            client_config: client_config.config().unwrap(),
+                            config_override,
+                        };
+                        let sut_layer = sut.config().unwrap();
+                        assert!(sut_layer
+                            .load::<#{SharedCredentialsCache}>()
+                            .is_some());
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                unitTest("test_overriding_credentials_cache_leads_to_shared_credentials_cache_in_layer") {
+                    rustTemplate(
+                        """
+                        use #{RuntimePlugin};
+
+                        // TODO(enableNewSmithyRuntimeCleanup): Should not need to provide a default once smithy-rs##2770
+                        //  is resolved
+                        let client_config = Config::builder()
+                            .credentials_provider(#{Credentials}::for_tests())
+                            .build();
+                        let config_override = Config::builder()
+                            .credentials_cache(#{CredentialsCache}::no_caching());
+                        let sut = ConfigOverrideRuntimePlugin {
+                            client_config: client_config.config().unwrap(),
+                            config_override,
+                        };
+                        let sut_layer = sut.config().unwrap();
+                        assert!(sut_layer
+                            .load::<#{SharedCredentialsCache}>()
+                            .is_some());
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                unitTest("test_overriding_cache_and_provider_leads_to_shared_credentials_cache_in_layer") {
+                    rustTemplate(
+                        """
+                        use #{RuntimePlugin};
+
+                        let client_config = Config::builder().build();
+                        let config_override = Config::builder()
+                            .credentials_cache(#{CredentialsCache}::lazy())
+                            .credentials_provider(#{Credentials}::for_tests());
+                        let sut = ConfigOverrideRuntimePlugin {
+                            client_config: client_config.config().unwrap(),
+                            config_override,
+                        };
+                        let sut_layer = sut.config().unwrap();
+                        assert!(sut_layer
+                            .load::<#{SharedCredentialsCache}>()
+                            .is_some());
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                unitTest("test_not_overriding_cache_and_provider_leads_to_no_shared_credentials_cache_in_layer") {
```

Comment:
Should there also be a test for:
```rust
                        let client_config = Config::builder()
                            .credentials_cache(#{CredentialsCache}::lazy())
                            .credentials_provider(#{Credentials}::for_tests())
                            .build();
                        let config_override = Config::builder().build();
```
?

What about:
```rust
                        let client_config = Config::builder()
                            .credentials_cache(#{CredentialsCache}::lazy())
                            .credentials_provider(#{Credentials}::for_tests())
                            .build();
                        let config_override = Config::builder()
                            .credentials_cache(#{CredentialsCache}::lazy())
                            .credentials_provider(<<<something else here>>>)
                            .build();
```
?

It should be possible to make a custom credentials provider and resolve credentials to verify it used the right one.

Reply:
Updated tests more thoroughly in 0e94e37.